ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/BeaconStateWithCache.java

```diff
@@ -172,6 +225,16 @@ public static BeaconStateWithCache fromBeaconState(BeaconState state) {
         state.getFinalized_checkpoint());
   }
 
+  public static TransitionCaches getTransitionCaches(BeaconState state) {
+    return state instanceof BeaconStateWithCache
+        ? ((BeaconStateWithCache) state).getTransitionCaches()
+        : TransitionCaches.getNoOp();
+  }
+
+  public TransitionCaches getTransitionCaches() {
+    return transitionCaches;
+  }
```

Comment:
Would it be better to add `getTransitionCaches` to `BeaconState` and then use polymorphism so `BeaconState` returns `TransitionCaches.getNoOp()` and `BeaconStateWithCache` overrides it to return a real cache?

That would avoid the `instanceof` checks here and in deepCopy.

Reply:
Not sure about this. I would keep the core `BeaconState` as clear as possible of any implementation specific stuff. I believe this central class would finally become pretty complex even without those 'tool' methods. 
To reduce boilerplate in the spec functions we could do like this https://github.com/Nashatyrev/artemis/pull/6 (though after formatting I'm not sure it looks much nicer)
However I have no strong opinion whether it should done this or that way.

Reply:
I'd stick with the current approach in this PR for now then.  https://github.com/Nashatyrev/artemis/pull/6 doesn't seem too worth it.

Let's see how it goes as we add more caches - we may come up with other ideas, especially if we do wind up moving to a more OO style as discussed above.